### ISSUE TITLE
Suggestion on builder

### ISSUE THREAD

**Issue by nickwph** [tomasbjerre/java-method-invocation-builder#1]
Hey! I like this project a lot and it a good idea to make retrofit like a builder.

However i would suggest to store the instance of class to invoke with in the front. This way I don't have to keep track of the instance until the method is used.

So.. instead of this
```
BitBucketServerServicePullRequestsBuilder.pullRequests().invoke(bitBucketServerService);
BitBucketServerServicePullRequestsBuilder.pullRequests().withAt("24").invoke(bitBucketServerService);
BitBucketServerServicePullRequestsBuilder.pullRequests().withAt("25").invoke(bitBucketServerService);
```

This would be much better.
```
BitBucketServerServicePullRequestsBuilder.with(bitBucketServerService).pullRequests().invoke();
BitBucketServerServicePullRequestsBuilder.with(bitBucketServerService).pullRequests().withAt("24").invoke();
BitBucketServerServicePullRequestsBuilder.with(bitBucketServerService).pullRequests().withAt("25").invoke();
```

Thanks!

**Comment by tomasbjerre** [comment]
I can see that there may be mistakes with transaction boundaries when the service is stored in the builder. Users may pass the builder as an argument to another service, with another transaction, and forget that the builder actually holds a reference to the service...

But that is just a speculation from my side, so I'm ok with adding this!

Perhaps `with(bitBucketServerService)` can be `on(bitBucketServerService)`?

And I think I would prefer this to be configured from annotations. So that you can generate builders that have the service in the `invoke`-method or that has the service in a `on`-method. To force developers into picking one pattern and stick with it.

Perhaps these alternatives?
```
@GenerateMethodInvocationBuilder(style = INVOKATION_STYLE.INVOKE_AT_END)
```
```
@GenerateMethodInvocationBuilder(style = INVOKATION_STYLE.INVOKE_WITH_ON)
```
```
@GenerateMethodInvocationBuilder(style = INVOKATION_STYLE.INVOKE_WITH_ON_OR_AT_END)
```

**Comment by nickwph** [comment]
Hey yes I totally agree with you on thread-safe matter. However I think invoking in the beginning would be a great feature, since I am used to use Android libraries `Glide` and `Picasa` that they also put `Context` first before the builder pattern. 

Anyways, passing builder to other classes is exactly what i want to do, because of convenience. 

Adding an attribute to configure the style is okay, however I don't understand what is `INVOKE_WITH_ON_OR_AT_END`?

Also, shouldn't `INVOKATION_STYLE` be `InvokationStyle` instead? Since it's a enum 

---

On the other hand, I think you could to save the service as `WeakReference`? in that case you can make sure you don't have a strong reference to the instance. 

```java
class GeneratedBuilder {

    WeakReference<Service> serviceRef;

    static GeneratedBuilder with(Service service) {     // or `on` if you prefer.
         serviceRef = new WeakReference<>(service);
    }

    Result invoke() {
        if (serviceRef.get() == null) {
            throw new ServiceDelocatedException();
        }
        // execute
    }  
}
```

**Comment by tomasbjerre** [comment]
Yes, `InvokationStyle` is correct.

I was thinking that `INVOKE_WITH_ON_OR_AT_END` can add both the on and the invoke(service) methods. To make the style optional for the user of the builder. But perhaps there can be a better name for that configuration?

**Comment by nickwph** [comment]
Haha I think you could use `AUTO` if you wanted to make it seem like the the annotation would figure out the style from inputs. But `INVOKE_WITH_ON_OR_AT_END` is okay too as long as there are documentations.

Anyways great job on this project! 😄

**Comment by tomasbjerre** [comment]
I think I have a working implementation now. Here are some examples:

https://github.com/tomasbjerre/java-method-invocation-builder/blob/master/example-maven/src/test/java/se/bjurr/jmib/ExamplesTest.java

What do you think?

**Comment by tomasbjerre** [comment]
I released this in 1.3. Open issue again if any problems.
